sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs

```diff
@@ -39,33 +40,80 @@ public sealed class PropertiesAccessCorrectField : PropertiesAccessCorrectFieldB
+                {
+                    foundField = ExtractFieldFromExpression(AccessorKind.Setter, assignment.Left, compilation);
+                }
+                else if (node is ArgumentSyntax argument && IsRefOrOut(argument.RefOrOutKeyword))
+                {
+                    foundField = ExtractFieldFromExpression(AccessorKind.Setter, argument.Expression, compilation);
+                }
+                if (foundField.HasValue && !assignments.ContainsKey(foundField.Value.Field))
+                {
+                    assignments.Add(foundField.Value.Field, foundField.Value);
+                }
+            }
+            return assignments.Values;
+
+            bool IsRefOrOut(SyntaxToken node) => node.IsKind(SyntaxKind.RefKeyword) || node.IsKind(SyntaxKind.OutKeyword);
         }
 
-        protected override FieldData? FindReturnedField(IPropertySymbol property, Compilation compilation)
+        protected override IEnumerable<FieldData> FindFieldReads(IPropertySymbol property, Compilation compilation)
         {
             // We don't handle properties with multiple returns that return different fields
-            if (property.GetMethod?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is AccessorDeclarationSyntax accessor &&
-                accessor.DescendantNodes().FirstOrDefault(n => n.RawKind == (int)SyntaxKind.ReturnStatement) is ReturnStatementSyntax returnStatement &&
-                returnStatement.Expression != null)
+            if (!(property.GetMethod?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is AccessorDeclarationSyntax getter))
             {
-                return ExtractFieldFromExpression(AccessorKind.Getter, returnStatement.Expression, compilation);
+                return Enumerable.Empty<FieldData>();
             }
-            return null;
+
+            var reads = new Dictionary<IFieldSymbol, FieldData>();
+            var notAssigned = getter.DescendantNodes().Select(n => n as ExpressionSyntax)
+                .WhereNotNull().Where(n => !IsLeftSideOfAssignment(n));
+            foreach (var expression in notAssigned)
+            {
+                var readField = ExtractFieldFromExpression(AccessorKind.Getter, expression, compilation);
+                // we only keep information for the first location of the symbol
+                if (readField.HasValue && !reads.ContainsKey(readField.Value.Field))
+                {
+                    reads.Add(readField.Value.Field, readField.Value);
+                }
+            }
+            return reads.Values;
+        }
+
+        protected override bool ShouldIgnoreAccessor(IMethodSymbol accessorMethod)
+        {
+            if (!(accessorMethod?.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() is AccessorDeclarationSyntax accessor))
+            {
+                // no accessor
+                return true;
+            }
+            if (accessor.DescendantNodes().Count(n => n is StatementSyntax) > 2)
```

Comment:
The code in lines 103-108 made my head hurt; I had to debug through the test to work out that the special case is "_ignore the accessor if it only contains throw_".

As it is, the code still raises an FP if the accessor is expression-bodied e.g.

```
private object baz2;
public object Baz2
{
    get /* 123 */ => /* 456 */ throw new System.InvalidOperationException(""); // Compliant
    set => throw new System.InvalidOperationException(""); // Compliant
}
```

The following code seems to work:
```
            // Special case: ignore the accessor if the only statement/expression is a throw.
            if (accessor.Body == null)
            {
                // Expression-bodied syntax
                return accessor.DescendantNodes().FirstOrDefault() is ArrowExpressionClauseSyntax arrowClause
                    && ShimLayer.CSharp.ThrowExpressionSyntaxWrapper.IsInstance(arrowClause.Expression);
            }

            // Statement-bodied syntax
            return (accessor.Body.DescendantNodes().Count(n => n is StatementSyntax) == 1 &&
                accessor.Body.DescendantNodes().Count(n => n is ThrowStatementSyntax) == 1);
```